src/crypto/ocb_openssl.cc

```diff
@@ -1,1206 +1,178 @@
-			#if (BPI == 8)
-			ad_checksum = xor_block(ad_checksum, ta[4]);
-			ad_checksum = xor_block(ad_checksum, ta[5]);
-			ad_checksum = xor_block(ad_checksum, ta[6]);
-			ad_checksum = xor_block(ad_checksum, ta[7]);
-			#endif
-			adp += BPI;
-		} while (--i);
-		ctx->ad_blocks_processed = ad_block_num;
-		ctx->ad_offset = ad_offset;
-		ctx->ad_checksum = ad_checksum;
-	}
-
-    if (final) {
-		block ta[BPI];
-
-        /* Process remaining associated data, compute its tag contribution */
-        remaining = ((unsigned)ad_len) % (BPI*16);
-        if (remaining) {
-			k=0;
-			#if (BPI == 8)
-			if (remaining >= 64) {
-				tmp.bl = xor_block(ad_offset, ctx->L[0]);
-				ta[0] = xor_block(tmp.bl, adp[0]);
-				tmp.bl = xor_block(tmp.bl, ctx->L[1]);
-				ta[1] = xor_block(tmp.bl, adp[1]);
-				ad_offset = xor_block(ad_offset, ctx->L[1]);
-				ta[2] = xor_block(ad_offset, adp[2]);
-				ad_offset = xor_block(ad_offset, ctx->L[2]);
-				ta[3] = xor_block(ad_offset, adp[3]);
-				remaining -= 64;
-				k=4;
-			}
-			#endif
-			if (remaining >= 32) {
-				ad_offset = xor_block(ad_offset, ctx->L[0]);
-				ta[k] = xor_block(ad_offset, adp[k]);
-				ad_offset = xor_block(ad_offset, getL(ctx, ntz(k+2)));
-				ta[k+1] = xor_block(ad_offset, adp[k+1]);
-				remaining -= 32;
-				k+=2;
-			}
-			if (remaining >= 16) {
-				ad_offset = xor_block(ad_offset, ctx->L[0]);
-				ta[k] = xor_block(ad_offset, adp[k]);
-				remaining = remaining - 16;
-				++k;
-			}
-			if (remaining) {
-				ad_offset = xor_block(ad_offset,ctx->Lstar);
-				tmp.bl = zero_block();
-				memcpy(tmp.u8, adp+k, remaining);
-				tmp.u8[remaining] = (unsigned char)0x80u;
-				ta[k] = xor_block(ad_offset, tmp.bl);
-				++k;
-			}
-			AES_ecb_encrypt_blks(ta,k,&ctx->encrypt_key);
-			switch (k) {
-				#if (BPI == 8)
-				case 8: ad_checksum = xor_block(ad_checksum, ta[7]);
-					/* fallthrough */
-				case 7: ad_checksum = xor_block(ad_checksum, ta[6]);
-					/* fallthrough */
-				case 6: ad_checksum = xor_block(ad_checksum, ta[5]);
-					/* fallthrough */
-				case 5: ad_checksum = xor_block(ad_checksum, ta[4]);
-					/* fallthrough */
-				#endif
-				case 4: ad_checksum = xor_block(ad_checksum, ta[3]);
-					/* fallthrough */
-				case 3: ad_checksum = xor_block(ad_checksum, ta[2]);
-					/* fallthrough */
-				case 2: ad_checksum = xor_block(ad_checksum, ta[1]);
-					/* fallthrough */
-				case 1: ad_checksum = xor_block(ad_checksum, ta[0]);
-			}
-			ctx->ad_checksum = ad_checksum;
-		}
-	}
+int ae_encrypt(ae_ctx *ctx, const void *nonce_ptr, const void *pt_ptr,
+	       int pt_len, const void *ad_ptr, int ad_len, void *ct_ptr,
+	       void *tag, int final) {
+  assert(final == AE_FINALIZE);
+  int len = 0;
+  int ciphertext_len = 0;
+  const unsigned char *nonce = (const unsigned char *)nonce_ptr;
+  const unsigned char *pt = (const unsigned char *)pt_ptr;
+  const unsigned char* ad = (const unsigned char *)ad_ptr;
+  unsigned char* ct = (unsigned char *)ct_ptr;
```

Comment:
Again, `reinterpret_cast`? (The same comment applies to the decryption function.)